src/server/encode.rs

```diff
@@ -1,279 +1,101 @@
-                self.dispatch(State::End, cx, buf)
-            } else {
-                // The response length lets us know if we are encoding
-                // our body in chunks or not
-                match self.res.len() {
-                    Some(body_len) => {
-                        self.body_len = body_len;
-                        self.dispatch(State::EncodeFixedBody, cx, buf)
-                    }
-                    None => self.dispatch(State::EncodeChunkedBody, cx, buf),
-                }
-            }
-        } else {
-            // If we haven't read the entire header it means `buf` isn't
-            // big enough. Break out of loop and return from `poll_read`
-            Poll::Ready(Ok(self.bytes_written))
-        }
-    }
-
-    /// Encode the body with a known length.
-    fn encode_fixed_body(
-        &mut self,
-        cx: &mut Context<'_>,
-        buf: &mut [u8],
-    ) -> Poll<io::Result<usize>> {
-        // Double check that we didn't somehow read more bytes than
-        // can fit in our buffer
-        debug_assert!(self.bytes_written <= buf.len());
-
-        // ensure we have at least room for 1 more byte in our buffer
-        if self.bytes_written == buf.len() {
-            return Poll::Ready(Ok(self.bytes_written));
-        }
-
-        // Figure out how many bytes we can read.
-        let upper_bound =
-            (self.bytes_written + self.body_len - self.body_bytes_written).min(buf.len());
-        // Read bytes from body
-        let range = self.bytes_written..upper_bound;
-        let inner_poll_result = Pin::new(&mut self.res).poll_read(cx, &mut buf[range]);
-        let new_body_bytes_written = match inner_poll_result {
-            Poll::Ready(Ok(n)) => n,
-            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
-            Poll::Pending => match self.bytes_written {
-                0 => return Poll::Pending,
-                n => return Poll::Ready(Ok(n)),
-            },
-        };
-        self.body_bytes_written += new_body_bytes_written;
-        self.bytes_written += new_body_bytes_written;
-
-        // Double check we did not read more body bytes than the total
-        // length of the body
-        debug_assert!(
-            self.body_bytes_written <= self.body_len,
-            "Too many bytes read. Expected: {}, read: {}",
-            self.body_len,
-            self.body_bytes_written
-        );
-
-        if self.body_len == self.body_bytes_written {
-            // If we've read the `len` number of bytes, end
-            self.dispatch(State::End, cx, buf)
-        } else if new_body_bytes_written == 0 {
-            // If we've reached unexpected EOF, end anyway
-            // TODO: do something?
-            self.dispatch(State::End, cx, buf)
-        } else {
-            // Else continue encoding
-            self.encode_fixed_body(cx, buf)
+        if self.response.header(DATE).is_none() {
+            let date = fmt_http_date(SystemTime::now());
```

Comment:
Not critical, but this should probably be cached, Node.js does it to the minute: https://github.com/nodejs/node/blob/4f0f2e713848817d49a4105a0736db2dfeed061e/lib/internal/http.js#L19-L29

(I may be willing to that in a follow-up pr.)